File: src/tiling/tiling.c
#include <ApplicationServices/ApplicationServices.h>
#include <Carbon/Carbon.h>

/**
 * Gets the bounds of the desktop, *excluding* the dock
 */
void getDesktopBounds(int* x, int* y, int* w, int* h) {
  HIRect bounds;
  HIWindowGetAvailablePositioningBounds(kCGNullDirectDisplay,kHICoordSpace72DPIGlobal,
    &bounds);

  *x = bounds.origin.x;
  *y = bounds.origin.y;
  *w = bounds.size.width;
  *h = bounds.size.height;
}

/**
 * Gets the frontmost app, which we can use to get the focused window
 */
static AXUIElementRef getFrontMostApp() {
  pid_t pid;
  ProcessSerialNumber psn;

  GetFrontProcess(&psn);
  GetProcessPID(&psn, &pid);

  return AXUIElementCreateApplication(pid);
}

/*
 * Moves the focused window to a specified location.
 */
void moveWindowImpl(int x, int y, int w, int h) {
  AXValueRef temp;
  AXUIElementRef frontMostApp;
  AXUIElementRef frontMostWindow;
  CGSize windowSize;
  CGPoint windowPosition;

  windowPosition.x = x;
  windowPosition.y = y;
  windowSize.width = w;
  windowSize.height = h;

  frontMostApp = getFrontMostApp();

  AXUIElementCopyAttributeValue(
    frontMostApp, kAXFocusedWindowAttribute, (CFTypeRef *)&frontMostWindow
  );

  temp = AXValueCreate(kAXValueCGPointType, &windowPosition);
  AXUIElementSetAttributeValue(frontMostWindow, kAXPositionAttribute, temp);
  CFRelease(temp);

  temp = AXValueCreate(kAXValueCGSizeType, &windowSize);
  AXUIElementSetAttributeValue(frontMostWindow, kAXSizeAttribute, temp);
  CFRelease(temp);
}

File: src/tiling/tiling.nim
{.compile: "tiling.c"}
proc getDesktopBounds(x: var int, y: var int, w: var int, h: var int) {.importc}
proc moveWindowImpl(x: int, y: int, w: int, h: int) {.importc}

proc tileWindow*(x: int, y: int, gap: int) =
  var dx = 0
  var dy = 0
  var dw = 0
  var dh = 0
  getDesktopBounds(dx, dy, dw, dh)

  let dr = dx + dw
  let db = dy + dh

  # Initialize the window location to the top left corner of the desktop
  var wx = dx
  var wy = dy
  var ww = 0
  var wh = 0

  if x == -1 or x == 1:
    ww = (dw / 2).int
  else:
    ww = dw

  if y == -1 or y == 1:
    wh = (dh / 2).int
  else:
    wh = dh

  wx = x * dw
  if wx < dx:
    wx = dx
  elif wx + ww > dr:
    wx = dr - ww

  wy = y * dh
  if wy < dy:
    wy = dy
  elif wy + wh > db:
    wy = db - wh

  wx += gap
  wy += gap

  ww -= gap * 2
  wh -= gap * 2

  moveWindowImpl(wx, wy, ww, wh)
